src/main/scala/strawman/collection/mutable/ArrayBuffer.scala

```diff
@@ -57,17 +62,145 @@ class ArrayBuffer[A] private (initElems: Array[AnyRef], initLength: Int)
   /** Overridden to use array copying for efficiency where possible. */
   override def ++[B >: A](xs: IterableOnce[B]): ArrayBuffer[B] = xs match {
     case xs: ArrayBuffer[B] =>
-      val elems = new Array[AnyRef](length + xs.length)
+      val elems = Array.ofDim[B](length + xs.length)(xs.elementClassTag)
       Array.copy(this.elems, this.start, elems, 0, this.length)
       Array.copy(xs.elems, xs.start, elems, this.length, xs.length)
-      new ArrayBuffer(elems, elems.length)
+      new ArrayBuffer[B](elems, elems.length)
     case _ => super.++(xs)
   }
 
+  /* overridden for specialization */
+  override def map[B](f: A => B): ArrayBuffer[B] = {
+    val length = this.length
+    val elems = this.elems
+    val start = this.start
+    var i = 0
+    // An alternative to using getSpecializedReturnType would be to map the first element unspecialized,
+    // use the runtime class of the resulting object for specialization, continue on the specialized path
+    // catching a possible ClassCastException to unspecialize later (by copying the already mapped values)
+    // if required. This would avoid the overhead of getSpecializedReturnType but it would also generate
+    // specialized collections from polymorphic mapping functions that happen to produce boxed primitive
+    // values.
+    val a = SpecializationUtil.getSpecializedReturnType(f).runtimeClass match {
+      case java.lang.Integer.TYPE =>
+        val ff = f.asInstanceOf[A => Int]
+        val a = new Array[Int](length)
+        while(i < length) {
+          a(i) = ff(elems(start + i))
+          i += 1
+        }
+        a
+      case java.lang.Long.TYPE =>
+        val ff = f.asInstanceOf[A => Long]
+        val a = new Array[Long](length)
+        while(i < length) {
+          a(i) = ff(elems(start + i))
+          i += 1
+        }
+        a
+      case java.lang.Double.TYPE =>
+        val ff = f.asInstanceOf[A => Double]
+        val a = new Array[Double](length)
+        while(i < length) {
+          a(i) = ff(elems(start + i))
+          i += 1
+        }
+        a
+      case _ =>
+        val ff = f.asInstanceOf[A => AnyRef]
+        val a = new Array[AnyRef](length)
+        while(i < length) {
+          // Note: Function1 is not specialized for AnyRef, so even if A is a primitive type
+          // we cannot call ff without boxing the argument.
```

Comment:
> we *could* pick out the other things that `Function1` is specialized for even if we can't store them that way, in order to save the boxing on the way in

Right, we could do this manually here. The bigger question is how far we want to go with primitive types. I only implemented specialization for Int, Long and Double, the same way that Java streams do. If we specialized for all types supported by Function1 Scala specialization would take care of this case.

> Any harm in leaving it as `A => Any`?

Then we need a cast in `a(i) = ff(elems(start + i))`. This would probably be the safer choice.